File: 1/2020.09.16/1021.Remove Outermost Parentheses/python/solution.py
class Solution:
    def removeOuterParentheses(self, S: str) -> str:
        count, res = 0, []
        for c in S:
            if c == '(' and count > 0:
                res.append(c)
            elif c == ')' and count > 1:
                res.append(c)
            count += 1 if c == '(' else -1
        return ''.join(res)

File: 1/2020.09.16/1021.Remove Outermost Parentheses/java/solution.java
class Solution {
    public String removeOuterParentheses(String S) {
        // using stack
        Stack<Character> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();
        for (char c : S.toCharArray()) {
            if (c == '(') {
                if (!stack.isEmpty()) {
                    sb.append(c);
                }
                stack.push(c);
            } else if (c == ')') {
                stack.pop();
                if (!stack.isEmpty()) {
                    sb.append(c);
                }
            }
        }
        return sb.toString();
    }

    public String removeOuterParentheses(String S) {
        // single pointer
        StringBuilder sb = new StringBuilder();
        int count = 0;
        for (char c : S.toCharArray()) {
            if (c == '(' && count++ > 0) {
                sb.append(c);
            } else if (c == ')' && count-- > 1) {
                sb.append(c);
            }
        }
        return sb.toString();
    }
}
